source/common/ssl/context_impl.cc

```diff
@@ -137,35 +147,53 @@ bssl::UniquePtr<SSL> ContextImpl::newSsl() const {
   return bssl::UniquePtr<SSL>(SSL_new(ctx_.get()));
 }
 
-bool ContextImpl::verifyPeer(SSL* ssl) const {
-  bool verified = true;
+int ContextImpl::verifyCallback(int ok, X509_STORE_CTX* store) {
+  if (X509_STORE_CTX_get_error_depth(store)) {
+    return ok;
+  }
 
+  X509* cert = X509_STORE_CTX_get_current_cert(store);
+  if (cert == nullptr) {
+    return ok;
+  }
+
+  SSL* ssl = reinterpret_cast<SSL*>(
+      X509_STORE_CTX_get_ex_data(store, SSL_get_ex_data_X509_STORE_CTX_idx()));
+  SSL_CTX* ctx = SSL_get_SSL_CTX(ssl);
+  ContextImpl* impl = reinterpret_cast<ContextImpl*>(SSL_CTX_get_app_data(ctx));
+  return impl->verifyCertificate(cert, ok);
+}
+
+int ContextImpl::verifyCertificate(X509* cert, int ok) {
```

Comment:
But then it would be inconsistent with `ContextImpl::verifySubjectAltName(X509* cert, ...)` and `ContextImpl::verifyCertificateHash(X509* cert, ...)`.